fdbserver/workloads/DiskThrottling.actor.cpp

```diff
@@ -28,102 +28,230 @@
-                           testDuration,
-                           Void());
-        } else
-            return Void();
-    }
-
-    Future<bool> check(Database const& cx) override { return true; }
-
-    void getMetrics(vector<PerfMetric>& m) override {}
-
-    static void checkDiskThrottleResult(Future<Void> res, WorkerInterface worker) {
-        if (res.isError()) {
-            auto err = res.getError();
-            if (err.code() == error_code_client_invalid_operation) {
-                TraceEvent(SevError, "ChaosDisabled")
-                    .detail("OnEndpoint", worker.waitFailure.getEndpoint().addresses.address.toString());
-            } else {
-                TraceEvent(SevError, "DiskThrottlingFailed")
-                    .detail("OnEndpoint", worker.waitFailure.getEndpoint().addresses.address.toString())
-                    .error(err);
-            }
-        }
-    }
-
-    ACTOR void doThrottle(WorkerInterface worker, double frequency, double minDelay, double maxDelay, double startDelay = 0.0) {
-        state Future<Void> res;
-        wait(::delay(startDelay));
-        SetFailureInjection::ThrottleDiskCommand throttleDisk;
-        throttleDisk.delayFrequency = frequency;
-        throttleDisk.delayMin = minDelay;
-        throttleDisk.delayMax = maxDelay;
-        SetFailureInjection req;
-        req.throttleDisk = throttleDisk;
-        res = worker.clientInterface.setFailureInjection.getReply(req);
-        wait(ready(res));
-        checkDiskThrottleResult(res, worker);
-    }
-
-    ACTOR static Future<Void> getAllWorkers(DiskThrottlingWorkload* self, std::vector<WorkerInterface>* result) {
-        result->clear();
-        std::vector<WorkerDetails> res =
-            wait(self->dbInfo->get().clusterInterface.getWorkers.getReply(GetWorkersRequest{}));
-        for (auto& worker : res) {
-            result->emplace_back(worker.interf);
-        }
-        return Void();
-    }
-
-    ACTOR static Future<Void> getAllStorageWorkers(Database cx, DiskThrottlingWorkload* self, std::vector<WorkerInterface>* result) {
-        result->clear();
-        state std::vector<WorkerInterface> res = wait(getStorageWorkers(cx, self->dbInfo, false));
-        for (auto& worker : res) {
-            result->emplace_back(worker);
-        }
-        return Void();
-    }
-
-    ACTOR template <class W>
-    Future<Void> throttleDiskClient(Database cx, DiskThrottlingWorkload* self) {
-        state double lastTime = now();
-        state double workloadEnd = now() + self->testDuration;
-        state std::vector<W> machines;
-        loop {
-            wait(poisson(&lastTime, 1));
-            wait(DiskThrottlingWorkload::getAllStorageWorkers(cx, self, &machines));
-            auto machine = deterministicRandom()->randomChoice(machines);
-            self->doThrottle(machine, self->throttleFrequency, self->throttleMin, self->throttleMax);
-        }
-    }
+	bool enabled;
```

Comment:
There is a lot of code duplication between this and throttling and bit flipping workloads, so it has a lot of the same issues.  

Additionally, I don't see why they need to be separate workloads.  Given that the command they send to the workers is a single command, `setFailureInjection`, it makes sense that there would be a single workload.   It could be configured via options to set one or both of the injection types.